File: src/AetherNet.Transport.WebRtc/WebRtcTransportService.cs
// SPDX-License-Identifier: MIT

using System.Collections.Concurrent;
using System.Diagnostics;
using AetherNet.Transport.Abstractions;
using AetherNet.Transport.Models;
using Microsoft.Extensions.Logging;
using SIPSorcery.Net;

namespace AetherNet.Transport.WebRtc;

/// <summary>
/// Direct peer-to-peer transport over a WebRTC <c>RTCDataChannel</c> (SIPSorcery, pure C#).
///
/// <para>NAT traversal is handled by ICE/STUN, with WebRTC's own TURN as last resort. The initial
/// SDP/ICE handshake is carried by an injected <see cref="IWebRtcSignaling"/> channel (e.g. the
/// AetherNet relay), so no central signalling server is required. Implements
/// <see cref="ITransportService"/> so <c>TransportManager</c> ranks it between the radio mesh
/// (cheap, proximity) and the QUIC/HTTP relay (last resort) — a direct internet path is used when
/// one can be negotiated, otherwise the relay carries the traffic.</para>
/// </summary>
public sealed class WebRtcTransportService : ITransportService, IAsyncDisposable
{
    private static readonly TimeSpan ConnectTimeout = TimeSpan.FromSeconds(20);

    private static readonly RTCIceServer[] DefaultIceServers =
    {
        new() { urls = "stun:stun.l.google.com:19302" },
    };

    private readonly string _localUhid;
    private readonly IWebRtcSignaling _signaling;
    private readonly List<RTCIceServer> _iceServers;
    private readonly ILogger<WebRtcTransportService>? _logger;
    private readonly PerTransportMetrics _metrics = new();
    private readonly ConcurrentDictionary<string, WebRtcPeerLink> _peers = new();
    private volatile bool _disposed;

    public WebRtcTransportService(
        string localUhid,
        IWebRtcSignaling signaling,
        IReadOnlyList<RTCIceServer>? iceServers = null,
        ILogger<WebRtcTransportService>? logger = null)
    {
        _localUhid = localUhid ?? throw new ArgumentNullException(nameof(localUhid));
        _signaling = signaling ?? throw new ArgumentNullException(nameof(signaling));
        _logger = logger;
        // null => the STUN default; an explicit (even empty) list is respected verbatim, so a
        // caller can pass an empty list to force host-candidate-only ICE (e.g. same-LAN / tests).
        _iceServers = (iceServers ?? DefaultIceServers).ToList();
        _signaling.SignalReceived += OnSignalReceived;
    }

    // ── ITransportService ──────────────────────────────────────────────
    public string Name => "WebRTC P2P";
    public bool IsAvailable => !_disposed;
    public long MaxBandwidthBps => 100_000_000;   // direct link — bounded by the local NIC
    public int MaxRangeMeters => 0;               // internet — unlimited
    public int PowerCostRelative => 45;           // between radio mesh (low) and relay (100)
    public int MaxConcurrentPeers => 256;
    public PerTransportMetrics? Metrics => _metrics;

    public event Action<string, byte[]>? DataReceived;

    public async Task<bool> SendAsync(string peerUhid, byte[] data, CancellationToken cancellationToken = default)
    {
        if (_disposed || string.IsNullOrEmpty(peerUhid)) return false;

        var link = await GetOrCreateLinkAsync(peerUhid, asInitiator: true, cancellationToken).ConfigureAwait(false);
        if (link is null) return false;

        var sw = Stopwatch.StartNew();
        var ok = await link.SendAsync(data, ConnectTimeout, cancellationToken).ConfigureAwait(false);
        sw.Stop();
        _metrics.RecordSample(sw.ElapsedMilliseconds, ok, ok ? data.Length : 0);
        return ok;
    }

    public async Task<bool> SendStreamAsync(string peerUhid, Stream stream, CancellationToken cancellationToken = default)
    {
        using var ms = new MemoryStream();
        await stream.CopyToAsync(ms, cancellationToken).ConfigureAwait(false);
        return await SendAsync(peerUhid, ms.ToArray(), cancellationToken).ConfigureAwait(false);
    }

    public bool IsConnected(string peerUhid) =>
        _peers.TryGetValue(peerUhid, out var link) && link.IsOpen;

    // ── Signalling inbound ─────────────────────────────────────────────
    private async void OnSignalReceived(WebRtcSignal signal)
    {
        if (_disposed || signal.ToUhid != _localUhid) return;
        try
        {
            switch (signal.Type)
            {
                case WebRtcSignalType.Offer:
                {
                    var link = await GetOrCreateLinkAsync(signal.FromUhid, asInitiator: false, CancellationToken.None)
                        .ConfigureAwait(false);
                    if (link is not null && signal.Sdp is not null)
                        await link.AcceptOfferAsync(signal.Sdp).ConfigureAwait(false);
                    break;
                }
                case WebRtcSignalType.Answer:
                    if (signal.Sdp is not null && _peers.TryGetValue(signal.FromUhid, out var answered))
                        answered.AcceptAnswer(signal.Sdp);
                    break;
                case WebRtcSignalType.IceCandidate:
                    if (_peers.TryGetValue(signal.FromUhid, out var cand))
                        cand.AddRemoteCandidate(signal);
                    break;
            }
        }
        catch (Exception ex)
        {
            _logger?.LogWarning(ex, "[WebRTC] signal handling failed from {Peer}", signal.FromUhid);
        }
    }

    private async Task<WebRtcPeerLink?> GetOrCreateLinkAsync(string peerUhid, bool asInitiator, CancellationToken ct)
    {
        if (_peers.TryGetValue(peerUhid, out var existing) && !existing.IsClosed)
        {
            if (asInitiator) await existing.WaitOpenAsync(ConnectTimeout, ct).ConfigureAwait(false);
            return existing;
        }

        var link = new WebRtcPeerLink(_localUhid, peerUhid, _iceServers, _signaling, OnPeerData, _logger);
        if (!_peers.TryAdd(peerUhid, link))
        {
            // Lost a race — discard ours, use the winner.
            await link.DisposeAsync().ConfigureAwait(false);
            _peers.TryGetValue(peerUhid, out var winner);
            return winner;
        }

        link.Closed += () => _peers.TryRemove(peerUhid, out _);
        await link.StartAsync(asInitiator).ConfigureAwait(false);

        if (asInitiator)
            await link.WaitOpenAsync(ConnectTimeout, ct).ConfigureAwait(false);

        return link;
    }

    private void OnPeerData(string peerUhid, byte[] data) => DataReceived?.Invoke(peerUhid, data);

    public async ValueTask DisposeAsync()
    {
        if (_disposed) return;
        _disposed = true;
        _signaling.SignalReceived -= OnSignalReceived;
        foreach (var link in _peers.Values)
            await link.DisposeAsync().ConfigureAwait(false);
        _peers.Clear();
        DataReceived = null;
    }
}

File: swift/Sources/AetherNetWebRTC/WebRtcTransportService.swift
// SPDX-License-Identifier: MIT

import Foundation
import CDataChannel
import AetherNetProtocol
import AetherNetWebRTCSignaling

/// Direct peer-to-peer transport over a WebRTC data channel (libdatachannel, portable C/C++).
///
/// NAT traversal is handled by ICE/STUN, with WebRTC's own TURN as last resort. The initial SDP/ICE
/// handshake is carried by an injected ``WebRtcSignaling`` channel (e.g. the AetherNet relay), so no
/// central signalling server is required. Conforms to `TransportService` so `rankTransports` slots
/// it between the radio mesh (cheap, proximity) and the QUIC/HTTP relay (last resort): a direct
/// internet path is used when one can be negotiated, otherwise the relay carries the traffic.
public final class WebRtcTransportService: TransportService, @unchecked Sendable {
    private static let connectTimeout: TimeInterval = 20

    /// Serverless default: NO ICE servers, so a node never contacts a STUN/TURN server. Direct
    /// links form on the same LAN or when a peer has a public address; for NAT traversal without a
    /// server, route through the circuit-relay-v2 transport (peers relay for peers). Opt into
    /// STUN/TURN by passing an explicit list.
    public static let defaultIceServers: [String] = []

    private let localUhid: String
    private let signaling: any WebRtcSignaling
    private let iceServers: [String]
    private let _metrics = PerTransportMetrics()

    private let lock = NSLock()
    private var peers: [String: WebRtcPeerLink] = [:]
    private var onData: (@Sendable (String, Data) -> Void)?
    private var disposed = false

    /// Creates the transport for `localUhid`, routing signalling through `signaling`.
    ///
    /// - Parameter iceServers: `nil` uses the serverless default of NO ICE servers
    ///   (host-candidate-only ICE; never contacts a STUN/TURN server, links form on the same LAN or
    ///   when a peer has a public address). For NAT traversal without a server, route through the
    ///   circuit-relay-v2 transport (peers relay for peers). An explicit list is respected verbatim,
    ///   so pass one to opt into STUN/TURN, or an empty list to keep host-candidate-only ICE
    ///   (same-LAN / tests). URLs are in libdatachannel form, e.g. `stun:host:port`,
    ///   `turn:user:pass@host:port`.
    public init(
        localUhid: String,
        signaling: any WebRtcSignaling,
        iceServers: [String]? = nil
    ) {
        precondition(!localUhid.isEmpty, "localUhid is required")
        self.localUhid = localUhid
        self.signaling = signaling
        self.iceServers = iceServers ?? Self.defaultIceServers

        let weakSelf = WeakBox(self)
        Task {
            await signaling.onSignal { signal in
                weakSelf.value?.handleSignal(signal)
            }
        }
    }

    /// Registers the handler for inbound bytes — the receive surface, mirroring
    /// `InProcessTransport.onDataReceived`.
    public func onDataReceived(_ handler: @escaping @Sendable (String, Data) -> Void) {
        lock.lock(); defer { lock.unlock() }
        onData = handler
    }

    // MARK: - TransportService metadata

    public var name: String { "WebRTC P2P" }

    public var isAvailable: Bool {
        lock.lock(); defer { lock.unlock() }
        return !disposed
    }

    public var maxBandwidthBps: Int64 { 100_000_000 }   // direct link — bounded by the local NIC
    public var maxRangeMeters: Int32 { 0 }              // internet — unbounded
    public var powerCostRelative: Int32 { 5 }          // between radio mesh (low) and relay (high)
    public var maxConcurrentPeers: Int32 { 256 }
    public var metrics: PerTransportMetrics? { _metrics }

    // MARK: - TransportService send / status

    public func sendAsync(
        peerUhid: String,
        data: Data,
        cancellationToken: CancellationToken?
    ) async -> Bool {
        lock.lock()
        let isDisposed = disposed
        lock.unlock()
        if isDisposed || peerUhid.isEmpty { return false }

        guard let link = await getOrCreateLink(peerUhid: peerUhid, asInitiator: true) else {
            return false
        }
        let start = DispatchTime.now()
        let ok = await link.send(data, openTimeout: Self.connectTimeout)
        let elapsedMs = Double(DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1e6
        _metrics.recordSample(rttMs: elapsedMs, success: ok, bytesTransferred: ok ? data.count : 0)
        return ok
    }

    public func sendStreamAsync(
        peerUhid: String,
        data: Data,
        cancellationToken: CancellationToken?
    ) async -> Bool {
        await sendAsync(peerUhid: peerUhid, data: data, cancellationToken: cancellationToken)
    }

    public func isConnected(peerUhid: String) -> Bool {
        lock.lock()
        let link = peers[peerUhid]
        lock.unlock()
        return link?.isOpen ?? false
    }

    /// Tears down all peer connections and stops accepting new ones.
    public func close() {
        lock.lock()
        if disposed { lock.unlock(); return }
        disposed = true
        let links = Array(peers.values)
        peers.removeAll()
        onData = nil
        lock.unlock()
        for link in links { link.close() }
    }

    // MARK: - Signalling inbound

    private func handleSignal(_ signal: WebRtcSignal) {
        lock.lock()
        let isDisposed = disposed
        lock.unlock()
        if isDisposed || signal.toUhid != localUhid { return }

        switch signal.type {
        case .offer:
            Task {
                if let link = await getOrCreateLink(peerUhid: signal.fromUhid, asInitiator: false),
                   let sdp = signal.sdp {
                    link.acceptOffer(sdp: sdp)
                }
            }
        case .answer:
            lock.lock()
            let link = peers[signal.fromUhid]
            lock.unlock()
            if let link, let sdp = signal.sdp {
                link.acceptAnswer(sdp: sdp)
            }
        case .iceCandidate:
            lock.lock()
            let link = peers[signal.fromUhid]
            lock.unlock()
            link?.addRemoteCandidate(signal)
        }
    }

    private func getOrCreateLink(peerUhid: String, asInitiator: Bool) async -> WebRtcPeerLink? {
        lock.lock()
        if disposed {
            lock.unlock()
            return nil
        }
        if let existing = peers[peerUhid], !existing.isClosed {
            lock.unlock()
            if asInitiator {
                _ = await existing.waitOpen(timeout: Self.connectTimeout)
            }
            return existing
        }

        let onData = self.onData
        let link = WebRtcPeerLink(
            localUhid: localUhid,
            peerUhid: peerUhid,
            iceServers: iceServers,
            signaling: signaling,
            onData: { peer, data in onData?(peer, data) })
        peers[peerUhid] = link
        lock.unlock()

        link.registerOnClosed { [weak self] in
            guard let self else { return }
            self.lock.lock()
            self.peers.removeValue(forKey: peerUhid)
            self.lock.unlock()
        }
        link.start(asInitiator: asInitiator)

        if asInitiator {
            _ = await link.waitOpen(timeout: Self.connectTimeout)
        }
        return link
    }
}

// MARK: - WeakBox

/// A `Sendable` weak reference holder, so a C/async callback can reach back to the transport without
/// extending its lifetime.
private final class WeakBox<T: AnyObject>: @unchecked Sendable {
    weak var value: T?
    init(_ value: T) { self.value = value }
}
